bfsaz/bfsaz.go

```diff
@@ -0,0 +1,414 @@
+	}, nil
+}
+
+func (b *bucket) stripPrefix(name string) string {
+	if b.config.Prefix == "" {
+		return name
+	}
+	name = strings.TrimPrefix(name, b.config.Prefix)
+	name = strings.TrimPrefix(name, "/")
+	return name
+}
+
+func (b *bucket) withPrefix(name string) string {
+	if b.config.Prefix == "" {
+		return name
+	}
+	return internal.WithinNamespace(b.config.Prefix, name)
+}
+
+// Glob implements bfs.Bucket.
+func (b *bucket) Glob(ctx context.Context, pattern string) (bfs.Iterator, error) {
+	// quick sanity check
+	if _, err := doublestar.Match(pattern, ""); err != nil {
+		return nil, err
+	}
+
+	return &iterator{
+		parent:  b,
+		ctx:     ctx,
+		pattern: pattern,
+	}, nil
+}
+
+// Head implements bfs.Bucket.
+func (b *bucket) Head(ctx context.Context, name string) (*bfs.MetaInfo, error) {
+	resp, err := b.NewBlockBlobURL(b.withPrefix(name)).
+		GetProperties(ctx, azblob.BlobAccessConditions{})
+	if err != nil {
+		return nil, normError(err)
+	}
+
+	return &bfs.MetaInfo{
+		Name:        name,
+		Size:        resp.ContentLength(),
+		ModTime:     resp.LastModified(),
+		ContentType: resp.ContentType(),
+		Metadata:    bfs.NormMetadata(transKeys(resp.NewMetadata(), "_", "-")),
+	}, nil
+}
+
+// Open implements bfs.Bucket.
+func (b *bucket) Open(ctx context.Context, name string) (bfs.Reader, error) {
+	resp, err := b.NewBlockBlobURL(b.withPrefix(name)).
+		Download(ctx, 0, azblob.CountToEnd, azblob.BlobAccessConditions{}, false)
+	if err != nil {
+		return nil, normError(err)
+	}
+
+	return &reader{
+		ReadCloser:    resp.Body(azblob.RetryReaderOptions{}),
+		ContentLength: resp.ContentLength(),
+	}, nil
+}
+
+// Create implements bfs.Bucket.
+func (b *bucket) Create(ctx context.Context, name string, opts *bfs.WriteOptions) (bfs.Writer, error) {
+	f, err := ioutil.TempFile("", "bfs-az")
+	if err != nil {
+		return nil, err
+	}
+
+	return &writer{
+		File: f,
+		ctx:  ctx,
+		blob: b.NewBlockBlobURL(b.withPrefix(name)),
+		opts: opts,
+	}, nil
+}
+
+// Remove implements bfs.Bucket.
+func (b *bucket) Remove(ctx context.Context, name string) error {
+	_, err := b.NewBlockBlobURL(b.withPrefix(name)).
+		Delete(ctx, "", azblob.BlobAccessConditions{})
+	if ne := normError(err); ne != nil && ne != bfs.ErrNotFound {
+		return ne
+	}
+	return nil
+}
+
+// Close implements bfs.Bucket.
+func (*bucket) Close() error { return nil }
+
+// --------------------------------------------------------------------
+
+func normError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	var se azblob.StorageError
+	if errors.As(err, &se) {
+		if se.ServiceCode() == azblob.ServiceCodeBlobNotFound {
+			return bfs.ErrNotFound
+		}
+		return err
+	}
+
+	if pe := pipeline.Cause(err); pe != nil {
+		if ue := errors.Unwrap(pe); ue != nil {
+			return ue
+		}
+		return pe
+	}
+
+	return err
+}
+
+func transKeys(meta map[string]string, from, to string) map[string]string {
+	for k, v := range meta {
+		if nk := strings.ReplaceAll(k, from, to); nk != k {
```

Comment:
This (replacing underscores/dashes) is coded in bfs.NormMetadata - code duplication?

Reply:
Or that's intended - no need to MIME-canonicalise keys?

Reply:
they are canonicalised in the next step